src/blog/articles/node-js-race-conditions/index.md

```diff
@@ -0,0 +1,351 @@
+
+One way to solve this problem is to isolate the 2 concurrent operations into _transactions_ and make sure that there is only one transaction running at a given time. This idea might look like this:
+
+{% responsiveImage './blog/articles/node-js-race-conditions/aurei-fixed-race-condition-node-js.png', 'Fixing a race condition using a transaction', { maxWidth: 848 }  %}
+
+In the last picture, we are using transactions to make sure that all the steps of **Component 1** happen in order before all the steps of **Component 2**. This avoids any _stale read_ and makes sure that every component always has an up to date view of the world before doing any change. You can stop holding your breath now, Julius!
+
+In the rest of this article, we will zoom in more on race conditions in the context of Node.js and we will see some other approaches that can allow dealing with them.
+
+
+## Can we have race conditions in Node.js?
+
+It is a common misconception to think that Node.js does not have race conditions because of its single-threaded nature. While it is true that in Node.js you would not have multiple threads competing for resources, you might still end up with events belonging to different logical transactions being executed in an order that might result in _stale reads_ and generate a race condition.
+
+In the example that we illustrated above, we intentionally represented the various events (_read_, _increase_ and _save_) as discrete units. Note how the system is never executing more than one event at the same time. This is a simple but accurate representation of how the Node.js event loop processes events on a single thread. Nonetheless, you can see that there might be situations where multiple logical transactions (e.g. multiple deposits) are scheduled concurrently on the event loop and the discrete events might end up being intermingled, which results in a race condition.
+
+So... **Yes**, we can have race conditions in Node.js!
+
+
+## A Node.js example with a race condition
+
+Now, let's talk some code! Let's try to re-create the Roman Empire simulation game example that we discussed above.
+
+In ancient Rome, Romans used to export olives and grapes (no wonders Italy is still famous worldwide for olive oil and wine!). In our game, we want to be able to sell olives and grapes as a mean to acquire more _aurei_. 
+
+We are going to have two functions that can increase the balance by `50` _aurei_ which we are going to call `sellOlives()` and `sellGrapes()`. We will also assume that every time the balance is changed, it is persisted to a data storage of sort (e.g. a database). For the sake of this example, we won't be implementing the data storage for real, but we will just simulate some random asynchronous delay before reading or modifying a global value. This will be enough to illustrate how we can end up with a race condition.
```

Comment:
I'd be cautious to mention a database here as an example. Databases have their own mean to solve race conditions and we should not try to encourage to solve those issues in the application. Probably a file or a webservice is a better example?